File: http.ts
import { CommandContext, makeCommand } from "./command_utils.ts";
import { ApplicationCommandOptionType } from "$discord-api-types";
import { sortBy } from "@std/collections";

enum HttpCommandOption {
  StatusCode = "status_code",
  Variant = "variant",
}

const httpVariants = sortBy([
  {
    label: "cat",
    buildUrl: (statusCode: number) => `https://http.cat/${statusCode}`,
  },
  {
    label: "mdn",
    buildUrl: (statusCode: number) =>
      `https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/${statusCode}`,
  },
  {
    label: "cat2",
    buildUrl: (statusCode: number) => `https://httpcats.com/${statusCode}.webp`,
  },
  {
    label: "dog",
    buildUrl: (statusCode: number) => `https://http.dog/${statusCode}.webp`,
  },
  {
    label: "goat",
    buildUrl: (statusCode: number) =>
      `https://httpgoats.com/${statusCode}.webp`,
  },
  {
    label: "duck",
    buildUrl: (statusCode: number) =>
      `https://httpducks.com/${statusCode}.webp`,
  },
  {
    label: "garden",
    buildUrl: (statusCode: number) => `https://http.garden/${statusCode}.webp`,
  },
  {
    label: "fish",
    buildUrl: (statusCode: number) => `https://http.fish/${statusCode}.webp`,
  },
  {
    label: "pizza",
    buildUrl: (statusCode: number) => `https://http.pizza/${statusCode}.webp`,
  },
], (v) => v.label);

const defaultVariant = httpVariants[0]!;

const httpCommandDefaultValue = {
  [HttpCommandOption.Variant]: defaultVariant.label,
};

export const httpCommand = makeCommand({
  name: "http",
  description: "Explain a HTTP status code with a thousand words",
  options: [
    {
      name: HttpCommandOption.StatusCode,
      description: "Your HTTP response status codes",
      type: ApplicationCommandOptionType.Integer,
      required: true,
      min_value: 100,
      max_value: 599,
    },
    {
      name: HttpCommandOption.Variant,
      description: `Determine the variant of the explanation. Default to ${
        httpCommandDefaultValue[HttpCommandOption.Variant]
      }`,
      type: ApplicationCommandOptionType.String,
      required: false,
      choices: httpVariants.map((v) => ({ name: v.label, value: v.label })),
    },
  ],
});

export function handleHttpCommand(
  { interactionData }: CommandContext,
) {
  const options = interactionData.options || [];

  const variantOption = options.find(
    (option) => option.name === HttpCommandOption.Variant,
  );
  const variant = variantOption?.type === ApplicationCommandOptionType.String
    ? variantOption.value
    : httpCommandDefaultValue[HttpCommandOption.Variant];

  const variantInfo = httpVariants.find((v) => v.label === variant) ||
    defaultVariant;

  const statusCodeOption = options.find(
    (option) => option.name === HttpCommandOption.StatusCode,
  );
  const statusCode =
    statusCodeOption?.type === ApplicationCommandOptionType.Integer
      ? statusCodeOption.value
      : 404;

  return {
    responseText: variantInfo.buildUrl(statusCode),
  };
}

File: commands/http.go
package commands

import "fmt"

type httpVariant struct {
	label    string
	buildURL func(statusCode int) string
}

// Sorted alphabetically by label — matches sortBy(v => v.label) in http.ts.
var httpVariants = []httpVariant{
	{"cat", func(c int) string { return fmt.Sprintf("https://http.cat/%d", c) }},
	{"cat2", func(c int) string { return fmt.Sprintf("https://httpcats.com/%d.webp", c) }},
	{"dog", func(c int) string { return fmt.Sprintf("https://http.dog/%d.webp", c) }},
	{"duck", func(c int) string { return fmt.Sprintf("https://httpducks.com/%d.webp", c) }},
	{"fish", func(c int) string { return fmt.Sprintf("https://http.fish/%d.webp", c) }},
	{"garden", func(c int) string { return fmt.Sprintf("https://http.garden/%d.webp", c) }},
	{"goat", func(c int) string { return fmt.Sprintf("https://httpgoats.com/%d.webp", c) }},
	{"mdn", func(c int) string {
		return fmt.Sprintf("https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/%d", c)
	}},
	{"pizza", func(c int) string { return fmt.Sprintf("https://http.pizza/%d.webp", c) }},
}

var defaultHTTPVariant = httpVariants[0] // "cat"

func handleHTTP(ctx CommandContext) (CommandResult, error) {
	statusCode := 404
	variant := defaultHTTPVariant

	for _, opt := range ctx.InteractionData.Options {
		switch opt.Name {
		case "status_code":
			statusCode = opt.IntValue()
		case "variant":
			v := opt.StringValue()
			for _, hv := range httpVariants {
				if hv.label == v {
					variant = hv
					break
				}
			}
		}
	}

	return CommandResult{ResponseText: variant.buildURL(statusCode)}, nil
}
